File: domain-persistence-modeling/java-spring-boot/src/main/java/x/scratch/child/ChildRecord.java
package x.scratch.child;

import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NonNull;
import org.springframework.data.annotation.Id;
import org.springframework.data.relational.core.mapping.Table;
import x.scratch.UpsertableRecord;

import java.util.Set;
import java.util.TreeSet;

import static lombok.AccessLevel.PACKAGE;

@AllArgsConstructor(access = PACKAGE)
@Builder(toBuilder = true)
@Data
@Table("child")
public class ChildRecord
        implements ChildSimpleDetails,
        UpsertableRecord<ChildRecord> {
    @Id
    private Long id;
    private @NonNull String naturalId;
    private String parentNaturalId;
    private String value;
    @Builder.Default
    private @NonNull Set<String> defaultSideValues = new TreeSet<>();
    @Builder.Default
    private @NonNull Set<String> sideValues = new TreeSet<>();
    private int version;

    static ChildRecord createRecordFor(final String naturalId) {
        return new ChildRecord(null, naturalId, null, null,
                new TreeSet<>(), new TreeSet<>(), 0);
    }

    @Override
    public ChildRecord upsertedWith(final ChildRecord upserted) {
        id = upserted.id;
        naturalId = upserted.naturalId;
        parentNaturalId = upserted.parentNaturalId;
        value = upserted.value;
        sideValues = new TreeSet<>(upserted.sideValues);
        version = upserted.version;
        return this;
    }
}

File: spring-boot-domain-persistence-modeling-kotlin/src/main/kotlin/x/domainpersistencemodeling/java/ChildRecord.kt
package x.domainpersistencemodeling.java

import org.springframework.data.annotation.Id
import org.springframework.data.relational.core.mapping.Table
import java.util.TreeSet

@Table("child")
data class ChildRecord(
        @Id var id: Long?,
        override var naturalId: String,
        override var parentNaturalId: String?,
        override var value: String?,
        override var subchildren: Set<String>,
        override var version: Int
) : ChildDetails,
        UpsertableRecord<ChildRecord> {
    override fun updateWith(upserted: ChildRecord): ChildRecord {
        id = upserted.id
        naturalId = upserted.naturalId
        parentNaturalId = upserted.parentNaturalId
        value = upserted.value
        subchildren = TreeSet(upserted.subchildren)
        version = upserted.version
        return this
    }

    companion object {
        internal fun createRecordFor(naturalId: String): ChildRecord {
            return ChildRecord(null, naturalId, null, null, TreeSet(),
                    0)
        }
    }
}
